### PR TITLE
Get rid of the dummies, and notice a few incorrect things

### PR DESCRIPTION
**Pull request by eggrobin**
`PileUp` serialization should be nicer now.

### CODE REVIEW

**Comment by eggrobin** [comment]
Edited to `ifdef` out some serialization that called the old signature of `Vessel::ReadFromMessage`; please go back to not breaking the build of `ksp_plugin`, because now I need to build in order to work on the C♯.

File: ksp_plugin/plugin.cpp [hunk 1]
```diff
@@ -386,6 +388,24 @@ void Plugin::InsertOrKeepVessel(GUID const& vessel_guid,
           << parent_index;
 }
 
+void Plugin::InsertUnloadedPart(
+    PartId part_id,
+    GUID const& vessel_guid,
+    RelativeDegreesOfFreedom<AliceSun> const& from_parent) {
+  not_null<Vessel*> vessel = find_vessel_by_guid_or_die(vessel_guid).get();
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: const

File: ksp_plugin/plugin.cpp [hunk 2]
```diff
@@ -386,6 +388,24 @@ void Plugin::InsertOrKeepVessel(GUID const& vessel_guid,
...
+  CHECK(is_new_unloaded(vessel));
+  RelativeDegreesOfFreedom<Barycentric> const relative =
+      PlanetariumRotation().Inverse()(from_parent);
+  ephemeris_->Prolong(current_time_);
+  AddPart(
+      vessel,
+      part_id,
+      1 * Kilogram,
+      vessel->parent()->current_degrees_of_freedom(current_time_) + relative);
+  not_null<Part*> part = vessel->part(part_id);
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: const

File: ksp_plugin/plugin.cpp [hunk 3]
```diff
@@ -386,6 +388,24 @@ void Plugin::InsertOrKeepVessel(GUID const& vessel_guid,
...
+    PartId part_id,
+    GUID const& vessel_guid,
+    RelativeDegreesOfFreedom<AliceSun> const& from_parent) {
+  not_null<Vessel*> vessel = find_vessel_by_guid_or_die(vessel_guid).get();
+  CHECK(is_new_unloaded(vessel));
+  RelativeDegreesOfFreedom<Barycentric> const relative =
+      PlanetariumRotation().Inverse()(from_parent);
+  ephemeris_->Prolong(current_time_);
+  AddPart(
+      vessel,
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: The indent is weird, I'd prefer to move this to the previous line and split the last parameter.

File: ksp_plugin/plugin.cpp [hunk 4]
```diff
@@ -1300,10 +1311,33 @@ std::uint64_t Plugin::FingerprintCelestialJacobiKeplerian(
...
 }
 
+void Plugin::AddPart(not_null<Vessel*> const vessel,
+                     PartId const part_id,
+                     Mass const mass,
+                     DegreesOfFreedom<Barycentric> const& degrees_of_freedom) {
+  std::map<PartId, not_null<Vessel*>>::iterator it;
+  bool emplaced;
+  std::tie(it, emplaced) = part_id_to_vessel_.emplace(part_id, vessel);
+  CHECK(emplaced);
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Print the part_id and the vessel guid.

File: ksp_plugin/plugin.cpp [hunk 5]
```diff
@@ -1300,10 +1311,33 @@ std::uint64_t Plugin::FingerprintCelestialJacobiKeplerian(
...
+  auto deletion_callback = [ it, &map = part_id_to_vessel_ ] {
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: No space before/after []

File: ksp_plugin/plugin.cpp [hunk 6]
```diff
@@ -1300,10 +1311,33 @@ std::uint64_t Plugin::FingerprintCelestialJacobiKeplerian(
...
+    map.erase(it);
+  };
+  auto part =
+      make_not_null_unique<Part>(part_id,
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Would fit on the previous line.

File: ksp_plugin/plugin.hpp [hunk 7]
```diff
@@ -149,16 +149,24 @@ class Plugin {
...
+  // |inserted=true|, |InsertUnloadedPart| must be called for its parts
+  // before the call to |AdvanceTime|, giving the vessel an initial state.
   // For a KSP |Vessel| |v|, the arguments correspond to |v.id|,
   // |v.orbit.referenceBody.flightGlobalsIndex|, |v.loaded|.
   virtual void InsertOrKeepVessel(GUID const& vessel_guid,
                                   Index parent_index,
                                   bool loaded,
                                   bool& inserted);
 
+  // Adds a part with the given part_id to the vessel with the given ID, which
```
Comment (pleroy) [inline ksp_plugin/plugin.hpp]: |part_id|
ID->guid

File: ksp_plugin/plugin.hpp [hunk 8]
```diff
@@ -427,8 +429,15 @@ class Plugin {
           physics::KeplerianElements<Barycentric>> const& keplerian_elements,
       MassiveBody const& body);
 
+  void AddPart(not_null<Vessel*> vessel,
```
Comment (pleroy) [inline ksp_plugin/plugin.hpp]: Comment: Adds a part to a vessel, recording it in the appropriate map and setting up a deletion callback.

File: ksp_plugin/vessel.hpp [hunk 9]
```diff
@@ -74,10 +67,9 @@ class Vessel {
   // |FreeParts|.
   virtual void KeepPart(PartId id);
   // Removes any part for which |AddPart| or |KeepPart| has not been called
-  // since the last call to |FreePart|, and removes the dummy part added by
-  // |InitializeUnloaded| if the latter has been called.  Checks that there are
-  // still parts left after the removals; thus a call to |AddParts| must occur
-  // before |FreeParts| is first called.
+  // since the last call to |FreePart|.  Checks that there are still parts left
+  // after the removals; thus a call to |AddParts| must occur before |FreeParts|
```
Comment (pleroy) [inline ksp_plugin/vessel.hpp]: AddPart (no s)

File: serialization/journal.proto [hunk 10]
```diff
@@ -774,6 +774,19 @@ message InsertOrKeepLoadedPart {
   optional In in = 1;
 }
 
+message InsertUnloadedPart {
+  extend Method {
+    optional InsertUnloadedPart extension = 5024;
```
Comment (pleroy) [inline serialization/journal.proto]: Invest in a new extension # to make sure that we don't confuse ourselves.

Comment (eggrobin) [inline serialization/journal.proto]: Done.

File: ksp_plugin/vessel.hpp (hunk 9 above)
Comment (eggrobin) [inline ksp_plugin/vessel.hpp]: Done.

File: ksp_plugin/plugin.hpp (hunk 8 above)
Comment (eggrobin) [inline ksp_plugin/plugin.hpp]: Done.

File: ksp_plugin/plugin.hpp (hunk 7 above)
Comment (eggrobin) [inline ksp_plugin/plugin.hpp]: Done.

File: ksp_plugin/plugin.cpp (hunk 6 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: Done.

File: ksp_plugin/plugin.cpp (hunk 5 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: Done; clang-format does that when capturing an expression by reference somehow O_o

File: ksp_plugin/plugin.cpp (hunk 4 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: I don't have the vessel guid.

File: ksp_plugin/plugin.cpp (hunk 3 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: Done.

File: ksp_plugin/plugin.cpp (hunk 1 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: Done.

**Pull request merged**